### PR TITLE
Optimize JSON Reader Whitespace Skipping with Vectorized Operations

### LINKED ISSUE #3353: Consider vectorizing whitespace skipping (ODL 8 supports .NET 8 or higher)
## Is your feature request related to a problem? Please describe.
Consider vectorizing whitespace skipping (ODL 8 supports .NET 8 or higher)

## Describe the solution you'd like
We could re-implement SkipWhitespaces(Async) to use IndexOfAnyExcept as follows:

```csharp
private bool SkipWhitespaces()
{
    while (true)
    {
        if (this.tokenStartIndex >= this.storedCharacterCount)
        {
            if (!this.ReadInput())
            {
                return false; // no more data
            }
        }

        ReadOnlySpan<char> span = this.characterBuffer.AsSpan(this.tokenStartIndex, this.storedCharacterCount - this.tokenStartIndex);

        // Find first non-whitespace (space, tab, CR, LF). Returns -1 if all are whitespace.
        int idx = span.IndexOfAnyExcept(' ', '\t', '\r', '\n');
        if (idx == -1)
        {
            // All whitespace so far; advance to end of current buffer and refill.
            this.tokenStartIndex = this.storedCharacterCount;
            continue;
        }

        this.tokenStartIndex += idx;
        return true;
    }
}
```

**NOTE**: To ensure this change delivers measurable improvements, benchmarking is essential to validate its impact on allocations and runtime performance. Possibly measure against (a) minified, (b) lightly spaced, (c) pretty-printed

### PR DESCRIPTION
**Pull request by WanjohiSammy**
### Issues

*This pull request fixes #3353.*

### Description

This PR improves perf of JSON parsing by implementing vectorized whitespace skipping operations using [ReadOnlySpan<char>.IndexOfAnyExcept](https://learn.microsoft.com/dotnet/api/system.memoryextensions.indexofanyexcept) and [SearchValues](https://learn.microsoft.com/dotnet/api/system.buffers.searchvalues-1).

#### Before
image

#### After
image

#### Benchmark project
[/WanjohiSammy/ODataJsonReaderBenchmarks-3](https://github.com/WanjohiSammy/ODataJsonReaderBenchmarks-3)

**Input data**:
https://github.com/WanjohiSammy/ODataJsonReaderBenchmarks-3/blob/main/ODataJsonReaderBenchmarks/SamplePayloadWithValues.json

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### Repository notes

Team members can start a CI build by adding a comment with the text `/AzurePipelines run` to a PR. A bot may respond indicating that there is no pipeline associated with the pull request. This can be ignored if the build is triggered.

Team members should **not** trigger a build this way for pull requests coming from forked repositories. They should instead trigger the build manually by setting the "branch" to `refs/pull/{prId}/merge` where `{prId}` is the ID of the PR.

### CODE REVIEW

File: src/Microsoft.OData.Core/Json/JsonReader.cs [hunk 1]
```diff
@@ -1491,21 +1518,87 @@ private void TryPopPropertyScope()
...
+                                   Vector.Equals(currentVector, carriageCharVector);
+
+                        // If not all characters are whitespace, find the first non-whitespace character
+                        if (!Vector.EqualsAll(isWhitespace, ushortMaxValueVector))
+                        {
+                            // Found at least one non-whitespace character in this vector
+                            for (int i = 0; i < vectorSize; i++)
+                            {
+                                // 0 means NOT whitespace
+                                if (isWhitespace[i] == ushort.MinValue)
```
Comment (habbes) [inline src/Microsoft.OData.Core/Json/JsonReader.cs]: What are the possible values for each `isWhiteSpace` element? Is each element either all 1 bits or all 0 bits? Or are there scenarios where some bits 0 and some are 1?

File: src/Microsoft.OData.Core/Json/JsonReader.cs [hunk 2]
```diff
@@ -1491,21 +1518,87 @@ private void TryPopPropertyScope()
...
         /// </summary>
-        /// <returns>true if a non-whitespace character was found in which case the tokenStartIndex is pointing at that character.
-        /// false if there are no non-whitespace characters left in the input.</returns>
+        /// <returns>true if a non-whitespace character was found; false if end of input was reached.</returns>
         private bool SkipWhitespaces()
         {
             do
             {
-                for (; this.tokenStartIndex < this.storedCharacterCount; this.tokenStartIndex++)
+                int remaining = this.storedCharacterCount - this.tokenStartIndex;
```
Comment (habbes) [inline src/Microsoft.OData.Core/Json/JsonReader.cs]: Is this method only called when there's a whitespace character detected?

File: src/Microsoft.OData.Core/Json/JsonReader.cs [hunk 3]
```diff
@@ -1491,21 +1518,87 @@ private void TryPopPropertyScope()
         }
 
         /// <summary>
-        /// Skips all whitespace characters in the input.
+        /// Vectorized implementation of whitespace skipping using Vector for cross-platform SIMD.
```
Comment (habbes) [inline src/Microsoft.OData.Core/Json/JsonReader.cs]: Do we have a different implementation of this method that is not vectorized? Why do we have to specify that it's vectorized/SIMD in the method summary? That's an implementation detail.

File: src/Microsoft.OData.Core/Json/JsonReader.cs (hunk 1 above)
Comment (WanjohiSammy) [reply src/Microsoft.OData.Core/Json/JsonReader.cs]: Each position in `isWhitespace ` will be `ushort.MaxValue(65535)` if the character is whitespace, or `0` otherwise.

Consider the following:
```cs
 Vector<ushort> isWhitespace = Vector.Equals(currentVector, spaceCharVector) |
                              Vector.Equals(currentVector, tabCharVector) |
                                  Vector.Equals(currentVector, newlineCharVector) |
                                  Vector.Equals(currentVector, carriageCharVector);
```

The vectors represent the whitespace characters ('', '\t', '\n', '\r') repeated across all the elements of the vector:
```cs
Vector<ushort> spaceCharVector = new Vector<ushort>(' '); // <32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32>
Vector<ushort> tabCharVector = new Vector<ushort>('\t'); // <9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9>
Vector<ushort> newlineCharVector = new Vector<ushort>('\n'); // <10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10>
Vector<ushort> carriageCharVector = new Vector<ushort>('\r'); // <13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13>
```

To check for whitespaces, we use `ushort.MaxValue` and `0`
- `ushortMaxValueVector` with `Vector.EqualsAll(...)`: All elements are `ushort.MaxValue` (used to check if all positions are whitespace).
- `zeroVector` with `Vector.EqualsAll(...)`: All elements are `0` (used to check if no positions are whitespace).
- `ushortMaxValueVector` with `Vector.EqualsAny(...)`: Checks if it contains whitespace
- `zeroVector` with `Vector.EqualsAny(...)`: Checks if it contains non-whitespace
```cs
Vector<ushort> ushortMaxValueVector = new Vector<ushort>(ushort.MaxValue);
Vector<ushort> zeroVector = new Vector<ushort>(0);
```

### Here is the full code
```cs
Vector<ushort> ushortMaxValueVector = new Vector<ushort>(ushort.MaxValue);
Vector<ushort> zeroVector = new Vector<ushort>(0);

var inputWithNoWhitespace = new string('a', 100);
var inputWithAllWhitespace = " \t\n\r \n           \t\n\r \t\n\n\n\n    \t\t\t\t\r\r\r\r     \r\r\n\r\r";
var inputWithSomeWhitespace = "Hello, \tWorld!\nThis is a test string.\r\nIt contains some whitespace characters.";

var currentVector = new Vector<ushort>(MemoryMarshal.Cast<char, ushort>(inputWithNoWhitespace.AsSpan(0, Vector<ushort>.Count)));
Vector<ushort> isWhitespaceWithNoWhitespace = GetWhitespaceMask(currentVector); // <0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>

Console.WriteLine(Vector.EqualsAll(isWhitespaceWithNoWhitespace, ushortMaxValueVector)); // False because there are no whitespace characters
Console.WriteLine(Vector.EqualsAll(isWhitespaceWithNoWhitespace, zeroVector)); // True because there are no whitespace characters

currentVector = new Vector<ushort>(MemoryMarshal.Cast<char, ushort>(inputWithAllWhitespace.AsSpan(0, Vector<ushort>.Count)));
Vector<ushort> isWhitespaceWithAllWhitespace = GetWhitespaceMask(currentVector); // <65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535>

Console.WriteLine(Vector.EqualsAll(isWhitespaceWithAllWhitespace, ushortMaxValueVector)); // True because all characters are whitespace
Console.WriteLine(Vector.EqualsAll(isWhitespaceWithAllWhitespace, zeroVector)); // False because all characters are whitespace

currentVector = new Vector<ushort>(MemoryMarshal.Cast<char, ushort>(inputWithSomeWhitespace.AsSpan(0, Vector<ushort>.Count)));
Vector<ushort> isWhitespaceWithSomeWhitespace = GetWhitespaceMask(currentVector); // <0, 0, 0, 0, 0, 0, 65535, 65535, 0, 0, 0, 0, 0, 0, 65535, 0>

Console.WriteLine(Vector.EqualsAll(isWhitespaceWithSomeWhitespace, ushortMaxValueVector)); // False because not all characters are whitespace
Console.WriteLine(Vector.EqualsAll(isWhitespaceWithSomeWhitespace, zeroVector)); // False because there are some whitespace characters

static Vector<ushort> GetWhitespaceMask(Vector<ushort> currentVector)
{
    Vector<ushort> spaceCharVector = new Vector<ushort>(' ');
    Vector<ushort> tabCharVector = new Vector<ushort>('\t');
    Vector<ushort> newlineCharVector = new Vector<ushort>('\n');
    Vector<ushort> carriageCharVector = new Vector<ushort>('\r');

    return Vector.Equals(currentVector, spaceCharVector) |
        Vector.Equals(currentVector, tabCharVector) |
        Vector.Equals(currentVector, newlineCharVector) |
        Vector.Equals(currentVector, carriageCharVector);
}
```

File: src/Microsoft.OData.Core/Json/JsonReader.cs (hunk 2 above)
Comment (WanjohiSammy) [reply src/Microsoft.OData.Core/Json/JsonReader.cs]: Not really. The method is just called regardless of whether there is whitespace or not. It returns `true` if the first non-whitespace is found and false otherwise.

File: src/Microsoft.OData.Core/Json/JsonReader.cs (hunk 3 above)
Comment (WanjohiSammy) [reply src/Microsoft.OData.Core/Json/JsonReader.cs]: I have updated it

File: src/Microsoft.OData.Core/Json/JsonReader.cs (hunk 1 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/Json/JsonReader.cs]: @WanjohiSammy Is there a reason you went for handwritten SIMD rather than the built-in, vectorized primitives approach? From my reading, `IndexOfAny`/`IndexOfAnyExcept` on spans are implemented with SIMD‑aware algorithms and often outperform handwritten loops, especially across CPUs with SSE2/AVX2/ARM NEON, without occasioning the challenge of maintaining this vectorization logic

File: test/UnitTests/Microsoft.OData.Core.Tests/Json/JsonReaderTests.cs [hunk 4]
```diff
@@ -240,6 +240,38 @@ public async Task ReadPrimitiveValueWithSpecialCharacters(string payload, string
...
+        {
+            using (var reader = new JsonReader(new StringReader(payload), isIeee754Compatible: false))
+            {
+                await reader.ReadAsync(); // Read start of object
+                await reader.ReadAsync(); // Read property name - Data
+                await reader.ReadAsync(); // Position reader at the beginning of string value
+                Assert.Equal(expected, await reader.GetValueAsync());
+            }
+        }
+
```
Comment (gathogojr) [inline test/UnitTests/Microsoft.OData.Core.Tests/Json/JsonReaderTests.cs:274]: It'd make sense to add tests that specifically validate vectorization doesn't change semantics for the following edge cases:
- Buffer boundary: `\r` at the end of buffer and `\n` at the beginning of the buffer - CRLF. The number of characters loaded into the buffer can be determined from the relevant `const` in `JsonReader` so it should be able to test this
- All-whitespace buffer: Long run of whitespace to stress the refill path and advance logic
- Different category of payloads - minified, lightly spaced, heavily spaced (We should do benchmarks around the same and show statistically significant speedups where whitespace is abundant; but show no regressions on minified payloads)

File: src/Microsoft.OData.Core/Json/JsonReader.cs [hunk 5]
```diff
@@ -1489,19 +1516,85 @@ private void TryPopPropertyScope()
...
-        /// <returns>true if a non-whitespace character was found in which case the tokenStartIndex is pointing at that character.
-        /// false if there are no non-whitespace characters left in the input.</returns>
+        /// <returns>true if a non-whitespace character was found; false if end of input was reached.</returns>
         private bool SkipWhitespaces()
         {
             do
             {
-                for (; this.tokenStartIndex < this.storedCharacterCount; this.tokenStartIndex++)
+                int remaining = this.storedCharacterCount - this.tokenStartIndex;
+                int vectorSize = Vector<ushort>.Count;
```
Comment (gathogojr) [inline src/Microsoft.OData.Core/Json/JsonReader.cs]: @WanjohiSammy My suggestion is to switch to `ReadOnlySpan<char>.IndexOfAnyExcept` with a cached `SearchValues<char>` of the 4 JSON whitespace chars (' ', '\t', '\r', '\n' - **only**). This is already vectorized by the runtime and keeps the code tiny. I think it should be at least as fast and much easier to maintain across x64/ARM64.

If custom vectorization logic should remain, add a small-span fast-path (e.g. < 16 or < 32 chars) to avoid paying setup cost on short inputs where SIMD can be slower than scalar

Again, if the custom handwritten logic should remain, factor the span-scan into one helper to avoid duplication and divergence between sync and async flavours - if possible.

File: test/UnitTests/Microsoft.OData.Core.Tests/Json/JsonReaderTests.cs (hunk 4 above)
Comment (WanjohiSammy) [reply test/UnitTests/Microsoft.OData.Core.Tests/Json/JsonReaderTests.cs:274]: I have added comprehensive tests using a variety of large string inputs with different combinations of whitespace characters (spaces, tabs, newlines, carriage returns) surrounding and within JSON string values. The test data includes cases such as:

```
public static TheoryData<string, string> LargeStringsWithWhitespace()
{
    return new TheoryData<string, string>
    {
        { new string(' ', 10000) + "\"HelloWorld\"" + new string(' ', 10000), "HelloWorld" },
        { new string('\t', 5000) + "\"TestString\"" + new string('\n', 5000), "TestString" },
        { new string('\r', 8000) + "\"WhitespaceTest\"" + new string(' ', 2000), "WhitespaceTest" },
        { new string(' ', 2500) + new string('\n', 2500) + "\"MixedWhitespace\"" + new string('\t', 2500) + new string('\r', 2500), "MixedWhitespace" },
        { new string(' ', 5) + new string('\n', 5) + "\"Hello\"" + new string(' ', 5) + new string('\t', 5) + "\"World\"" + new string('\r', 5) + new string(' ', 5) + new string('\n', 5) + "!", "Hello" },
        { new string(' ', 5) + new string('\n', 5) + "\"Hello" + new string(' ', 5) + new string('\t', 5) + "World" + new string('\r', 5) + new string(' ', 5) + new string('\n', 5) + "!\"", "Hello     \t\t\t\t\tWorld\r\r\r\r\r     \n\n\n\n\n!" },
    };
}
```

File: src/Microsoft.OData.Core/Json/JsonReader.cs [hunk 6]
```diff
@@ -2382,21 +2478,187 @@ static async ValueTask<bool> AwaitReadInputAsync(JsonReader thisParam, ValueTask
                         return false; // EOF
                     }
 
-                    while (thisParam.tokenStartIndex < thisParam.storedCharacterCount)
+                    int remaining = thisParam.storedCharacterCount - thisParam.tokenStartIndex;
+                    if (remaining <= 0)
+                    {
+                        // All characters were whitespace, need to read more
+                        thisParam.tokenStartIndex = thisParam.storedCharacterCount;
```
[the remark this answers is not in the document]
Comment (WanjohiSammy) [reply src/Microsoft.OData.Core/Json/JsonReader.cs]: This not applicable due to refactoring

File: src/Microsoft.OData.Core/Json/JsonReader.cs [hunk 7]
```diff
@@ -1491,21 +1518,93 @@ private void TryPopPropertyScope()
...
+
+                                    // Found non-whitespace character
+                                    return true; 
+                                }
+                            }
+
+                            // This should never happen due to the EqualsAll check above
+                            // But if it does, advance and return
+                            this.tokenStartIndex += (processedVectors + 1) * vectorSize;
+                            return true;
```
[the remark this answers is not in the document]
Comment (WanjohiSammy) [reply src/Microsoft.OData.Core/Json/JsonReader.cs]: This not applicable due to refactoring

File: src/Microsoft.OData.Core/Json/JsonReader.cs (hunk 1 above)
Comment (WanjohiSammy) [reply src/Microsoft.OData.Core/Json/JsonReader.cs]: I have refactored the whitespace skipping logic to eliminate manual vectorization and replaced it with the recommended approach using `SearchValues` and `IndexOfAnyExcept`. This simplifies the code, improves maintainability, and leverages efficient runtime vectorization.

File: src/Microsoft.OData.Core/Json/JsonReader.cs (hunk 5 above)
Comment (WanjohiSammy) [reply src/Microsoft.OData.Core/Json/JsonReader.cs]: I have incorporated the recommended solution:
https://github.com/OData/odata.net/pull/3381#discussion_r2455168531

**Review by gathogojr — DISMISSED** [review]
Left a comment, otherwise LGTM

**Comment by gathogojr** [comment]
@WanjohiSammy Given our update strategy, it's better to use **Update with rebase** than **Update with merge commit** - that is usually selected by default, and what you appear to use often:

image

**Review by xuzhg — DISMISSED** [review]

**Comment by WanjohiSammy** [comment]
> Given our update strategy, it's better to use **Update with rebase** than **Update with merge commit** - that is usually selected by default, and what you appear to use often:
> 
@gathogojr This was not possible on the GitHub because of diff conflicts. I had to resolve this manually(locally) and rebase

**Review by gathogojr — APPROVED** [review]
